src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp

```diff
@@ -0,0 +1,188 @@
+
+/// Recursively computes a set of root object expressions for \p expr.
+static void root_objects_rec(
+  const exprt &expr,
+  const namespacet &ns,
+  messaget log,
+  std::unordered_set<exprt, irep_hash> &dest)
+{
+  if(expr.id() == ID_symbol)
+  {
+    dest.insert(expr);
+  }
+  else if(expr.id() == ID_index)
+  {
+    root_objects_rec(to_index_expr(expr).array(), ns, log, dest);
+  }
+  else if(expr.id() == ID_member)
+  {
+    const typet &type = to_member_expr(expr).struct_op().type();
+    if(
+      type.id() == ID_struct || type.id() == ID_struct_tag ||
+      type.id() == ID_union || type.id() == ID_union_tag)
+    {
+      root_objects_rec(to_member_expr(expr).compound(), ns, log, dest);
+    }
+    else
+    {
+      throw unsupported_operation_exceptiont(
+        "unexpected assignment to member of '" + type.id_string() + "'");
+    }
+  }
+  else if(expr.id() == ID_if)
+  {
+    root_objects_rec(to_if_expr(expr).true_case(), ns, log, dest);
+    root_objects_rec(to_if_expr(expr).false_case(), ns, log, dest);
+  }
+  else if(expr.id() == ID_typecast)
+  {
+    root_objects_rec(to_typecast_expr(expr).op(), ns, log, dest);
+  }
+  else if(
+    expr.id() == ID_byte_extract_little_endian ||
+    expr.id() == ID_byte_extract_big_endian)
+  {
+    root_objects_rec(to_byte_extract_expr(expr).op(), ns, log, dest);
+  }
+  else if(expr.id() == ID_complex_real)
+  {
+    root_objects_rec(to_complex_real_expr(expr).op(), ns, log, dest);
+  }
+  else if(expr.id() == ID_complex_imag)
+  {
+    root_objects_rec(to_complex_imag_expr(expr).op(), ns, log, dest);
+  }
+  else if(
+    const auto &deref =
+      expr_try_dynamic_cast<dereference_exprt>(skip_typecast(expr)))
+  {
+    // normalise the dereferenced pointer into `pointer + offset`, extract
+    // pointer expression and try some simplifications
+    const exprt &pointer = pointer_arithmetict(deref->pointer()).pointer;
+    const source_locationt source_location = expr.source_location();
+    if(const auto &if_expr = expr_try_dynamic_cast<if_exprt>(pointer))
+    {
+      // split on disjunctions
+      // *(c ? true_case : false_case) => rec(*true_case); rec(*false_case)
+      dereference_exprt if_true(if_expr->true_case());
+      if_true.add_source_location() = source_location;
+      root_objects_rec(if_true, ns, log, dest);
+      dereference_exprt if_false(if_expr->false_case());
+      if_false.add_source_location() = source_location;
+      root_objects_rec(if_false, ns, log, dest);
+    }
+    else if(
+      const auto &address_of_expr =
+        expr_try_dynamic_cast<address_of_exprt>(pointer))
+    {
+      const exprt &object = skip_typecast(address_of_expr->object());
+      if(const auto &index = expr_try_dynamic_cast<index_exprt>(object))
+      {
+        // *(&arr[idx]) => rec(arr)
+        const exprt &arr = index->array();
+        if(arr.type().id() == ID_array)
+        {
+          root_objects_rec(arr, ns, log, dest);
+        }
+      }
```

Comment:
Is this needed? You'd arrive at the same result via the `else` branch (and some recursive calls).

Reply:
fixed